avalon/nuke/lib.py

```diff
@@ -3,82 +3,283 @@
+def select_nodes(nodes):
+    """Selects all inputed nodes
+
+    Arguments:
+        nodes (list): nuke nodes to be selected
     """
+    assert isinstance(nodes, (list, tuple)), "nodes has to be list or tuple"
+
+    [n['selected'].setValue(True) for n in nodes]
+
+
+def add_publish_knob(node):
+    '''Adds Publish node to node
+
+    Arguments:
+        node (obj): nuke node to be processed
+
+    Returns:
+        node (obj): processed nuke node
+    '''
+    if "publish" not in node.knobs():
+        divider = nuke.Text_Knob('')
+        knob = nuke.Boolean_Knob("publish", "Publish", True)
+        knob.setFlag(0x1000)
+        node.addKnob(divider)
+        node.addKnob(knob)
+    return node
+
+
+def set_avalon_knob_data(node, data={}, prefix="ak:"):
+    """ Sets a data into nodes's avalon knob
+
+    Arguments:
+        node (obj): Nuke node to imprint with data,
+        data ( dict): Data to be imprinted into AvalonTab
+        prefix (str, optional): filtering prefix
+
+    Returns:
+        node (obj)
+
+    Examples:
+        data = {
+            'asset': 'sq020sh0280',
+            'family': 'render',
+            'subset': 'subsetMain'
+        }
+    """
+    knobs = [
+        {"name": 'AvalonTab', "value": '', "type": "Tab_Knob"},
+        {"name": 'begin', "value": 'Avalon data group',
+         "type": "Tab_Knob", "group": 2},
+        {"name": '__divider__'},
+        {"name": 'avalon_data',
+         "value": 'Warning! Do not change following data!',
+         "type": "Text_Knob"},
+        {"name": '__divider__'},
+        {"name": 'begin',
+         "value": 'Avalon data group',
+         "type": "Tab_Knob", "group": -1}
+    ]
+    non_hiden = ["asset", "subset", "name", "namespace"]
+
     try:
-        node['avalon'].value()
-    except Exception:
-        tab = nuke.Tab_Knob("Avalon")
-        uk = nuke.Text_Knob('avalon', 'avalon data')
-        node.addKnob(tab)
-        node.addKnob(uk)
-        log.info("created new user knob avalon")
+        # create Avalon Tab and basic knobs
+        for k in knobs[:-1]:
+            if not k.get("group"):
+                if "__divider__" in k["name"]:
+                    knob = nuke.Text_Knob("")
+                    node.addKnob(knob)
+                elif k["name"] not in node.knobs().keys():
+                    knob = eval("nuke.{type}('{name}')".format(**k))
+                    node.addKnob(knob)
+                    try:
+                        knob.setValue(k['value'])
+                    except TypeError as E:
+                        print(E)
+            else:
+                if k["name"] not in node.knobs().keys():
+                    knob = eval((
+                        "nuke.{type}('{name}',"
+                        "'{value}', {group})"
+                    ).format(**k))
+                    node.addKnob(knob)
+
+        # add avalon knobs for imprinting data
+        for k, v in data.items():
+            label = k
```

Comment:
If you are setting `label = k` and value = `str(v)` anyway , why not do `for key, value in data.items()` and just do `value = str(value)`?